### ISSUE TITLE
Batched training pipeline + basis differentiation: match ParametricDFT.jl main

### ISSUE THREAD

**Issue by zazabap** [zazabap/pdft#5]
## Context

PR #4 lands the GPU benchmark harness. Real GPU `light` runs (n_train=5–10, epochs=100, lr=0.01) on both QuickDraw (m=n=5) and DIV2K-8q (m=n=8) surfaced two related problems:

1. **All four quantum bases under-train** at the `light` preset relative to classical baselines (PSNR ~10–18 dB vs FFT ~16–32 dB).
2. **QFT and EntangledQFT produce bit-identical metrics**, and TEBD / MERA collapse to a second bit-identical pair. Across 5 different test images at 4 keep ratios, every digit matches:

   ```
   QuickDraw light:                         DIV2K-8q light:
     qft           PSNR 10.45 10.81 ...       qft           PSNR 16.97 17.49 17.78 17.98
     entangled_qft PSNR 10.45 10.81 ...       entangled_qft PSNR 16.97 17.49 17.78 17.98   <-- identical
     tebd          PSNR 11.66 12.06 ...       tebd          PSNR 17.05 17.40 ...
     mera          (skipped)                  mera          PSNR 17.05 17.40 ...           <-- identical
   ```

The two issues are linked: the training pipeline drops the bases in a regime where their symmetries don't break.

---

## (1) Batched training pipeline matching `ParametricDFT.jl` main

`pdft.train_basis` currently mirrors only the **inner** optimization path of upstream `ParametricDFT.jl`, by design (`src/pdft/training.py` docstring: \"Phase 1 does not port batches, epochs, validation splits, LR schedules, or checkpointing\"). Upstream `main` actually exposes the full pipeline:

```julia
# ParametricDFT.jl src/training.jl::_train_basis_core (main branch)
function _train_basis_core(
    dataset::Vector{<:AbstractMatrix}, ...;
    epochs::Int,
    batch_size::Int = 1,
    validation_split::Float64,
    early_stopping_patience::Int,
    optimizer = :adam,
    device = :cpu,
    warmup_frac::Float64 = 0.05,
    lr_peak::Float64 = 0.01,
    lr_final::Float64 = 0.001,
    max_grad_norm::Union{Nothing, Float64} = nothing,
)
```

The benchmark fork (`ParametricDFT-Benchmarks.jl/config.jl`) calls into this pipeline with concrete settings:

| preset | epochs | n_train | batch_size | optimizer | warmup_frac | lr_peak | lr_final | max_grad_norm |
| --- | ---: | ---: | ---: | --- | ---: | ---: | ---: | ---: |
| smoke | 2 | 5 | 16 | adam | 0.05 | 0.01 | 0.001 | 1.0 |
| light | 5 | 10 | 8 | adam | 0.05 | 0.01 | 0.001 | 1.0 |
| moderate | 10 | 20 | 16 | adam | 0.05 | 0.01 | 0.001 | 1.0 |
| heavy | 20 | 50 | 16 | adam | 0.05 | 0.01 | 0.001 | 1.0 |
| generalized | 40 | 500 | 50 | adam | 0.05 | 0.003 | 0.0003 | 1.0 |

Note: in the Julia benchmarks `epochs` means full passes over the dataset (each pass = `n_train / batch_size` optimizer steps), giving 540 optimizer steps for the `generalized` paper preset — vs. our current Python `light` which does 100 single-target steps starting fresh per image.

### What to add to `pdft`

A new `pdft.train_basis_batched(basis, *, dataset, epochs, batch_size, optimizer, validation_split=0.0, early_stopping_patience=None, lr_peak, lr_final, warmup_frac=0.05, max_grad_norm=None, device=\"cpu\", seed=0) -> TrainingResult` that:

- Cycles a single `basis` through the full dataset `epochs` times (one shared basis, not fresh-per-image as the current P-pairing harness does).
- Per step: gather a `batch_size` slice, compute the mean loss and its gradient, apply Riemannian update with the cosine-with-warmup learning rate, optionally clip gradients to `max_grad_norm`.
- Validation split + patience-based early stopping (optional; `validation_split=0.0` disables).
- Optionally emit per-step loss + per-epoch (train + val) loss history.

This is the deferred \"Phase 2\" work referenced in `src/pdft/training.py` and CLAUDE.md (\"Batched training is open work in #2\").

### Acceptance for (1)

- [ ] Public `train_basis_batched` API, JAX pytree-generic across QFT/EntangledQFT/TEBD/MERA.
- [ ] Cosine-with-warmup schedule reproducible by tests against the closed-form formula in `ParametricDFT.jl/src/training.jl::_cosine_with_warmup`.
- [ ] Batched gradient: parametric tests showing `batch_size=1` reproduces single-target results (parity with current `train_basis`).
- [ ] Gradient clipping verified by injecting a synthetic explosion.
- [ ] benchmarks/config.py gains a parallel `BATCHED_PRESETS_*` set (or replaces existing) with the table above; benchmarks/run_*.py switches to one shared basis per dataset (no longer P-pairing).
- [ ] benchmarks/evaluation.evaluate_basis_per_image becomes evaluate_basis_shared (one trained basis evaluated against all `n_test` images).
- [ ] DIV2K-8q `moderate` smoke with batch_size=64 completes on 24 GB RTX 3090 within wall-clock budget (target < 1 h).

---

## (2) Basis differentiation — why QFT == EntangledQFT and TEBD == MERA

Investigation under PR #4. Each basis class has its own `forward_transform` and `inverse_transform`, but they collapse into pairs at `light` preset because of three compounding factors:

1. **Identical analytic initialization for some pairs.** `QFTBasis.__init__` uses `qft_code(m, n)` to seed `tensors`, producing the exact analytic QFT. `EntangledQFTBasis.__init__` is QFT + an entanglement layer whose 8 controlled-phase gates are seeded with **identity-equivalent phases** (each ≈ 1, so the layer acts as identity at init). Mathematically `EntangledQFT_init(image) == QFT_init(image)`. Only the trained drift breaks this, and at lr=0.01 with 100 GD steps from the same starting forward output, both bases converge to the same neighborhood.

2. **`pdft.train_basis` only updates forward `tensors`, not `inv_tensors`** (verified at `src/pdft/training.py:79–82`). `recover` calls `basis.inverse_transform(...)` which uses the **untrained** `inv_tensors`. For all four basis classes, `inv_tensors` is initialized to the analytic inverse circuit — and these analytic inverses produce the same operation when the forward has not differentiated. In practice the post-training forward output is filtered through an inverse circuit that is identical for QFT and EntangledQFT (analytic IQFT), giving identical reconstructions.

3. **Single-target training (P-pairing) destabilizes the low-frequency basis distinctions.** Each basis class is trained separately on the same image with no shared signal across images. The trained drift in `tensors` is dominated by a per-image L1 spike that's the same for QFT and EntangledQFT because both start from the same point.

A quick check that would confirm (1): print `np.allclose(QFTBasis(m=8, n=8).tensors[i], EntangledQFTBasis(m=8, n=8).tensors[i])` for the `i` indices that the entanglement layer should differ at; if True, the layer initializes as identity. Same for TEBD/MERA — check whether `TEBDBasis(m=8, n=8).inv_tensors == MERABasis(m=8, n=8).inv_tensors` (after pytree-flattening and matching shapes).

### What to fix

- **Random initialization** (or perturbation off the analytic identity) for the EntangledQFT entanglement-layer phases, TEBD CP gates, and MERA gates. Optional `init_scale: float = 0.01` parameter on each basis class to add Gaussian noise.
- **Train both `tensors` and `inv_tensors`** if the basis is being used in a compress→recover round-trip, OR document the asymmetry explicitly and test that `inverse_transform` parity holds. The cleanest fix: have the optimizer drive `(tensors, inv_tensors)` jointly to minimize the round-trip reconstruction loss, not just the forward L1.
- **Re-test with batched training (issue (1))**: with `batch_size=64, epochs=20+`, the bases should differentiate even under the current `tensors`-only training, because the cumulative drift per gate becomes meaningful.

### Acceptance for (2)

- [ ] At m=n=5 with batched training (preset moderate/heavy), all four basis classes produce **distinguishable** mean-PSNR at every keep ratio on a 50-image QuickDraw test split (no two pairs match to more than 2 decimal places).
- [ ] At m=n=8 on DIV2K-8q (preset moderate), QFT < EntangledQFT in MSE at keep_ratio=0.05, demonstrating the entanglement layer learns to do something useful.
- [ ] A unit test in `tests/test_basis.py`: `np.allclose(...)` between QFTBasis and EntangledQFTBasis tensor lists at default init **fails** (after random-init or perturbation is in place). Same for TEBD/MERA.

---

## Cross-references

- Spec: `docs/superpowers/specs/2026-04-25-pdft-gpu-benchmarks-design.md` §1 (out of scope: \"Julia-equivalent batched / scheduled training\").
- Plan: `docs/superpowers/plans/2026-04-25-pdft-gpu-benchmarks.md` §8 (future work).
- Upstream Julia: `ParametricDFT.jl/src/training.jl::_train_basis_core` (main branch, commit `a201a27e47df2f0f3ab460f83d49b6e5f5d1e9ef`).
- Benchmark fork: `ParametricDFT-Benchmarks.jl/config.jl` for the preset table.
- Current pdft Phase-1 limit: `src/pdft/training.py` module docstring.

🤖 Filed during /loop session after PR #4
